src/core/recv_buffer.c

```diff
@@ -170,10 +205,52 @@ QuicRecvBufferIncreaseVirtualBufferLength(
     _In_ uint32_t NewLength
     )
 {
+    CXPLAT_DBG_ASSERT(RecvBuffer->RecvMode != QUIC_RECV_BUF_MODE_APP_OWNED);
     CXPLAT_DBG_ASSERT(NewLength >= RecvBuffer->VirtualBufferLength); // Don't support decrease.
     RecvBuffer->VirtualBufferLength = NewLength;
 }
 
+_IRQL_requires_max_(DISPATCH_LEVEL)
+QUIC_STATUS
+QuicRecvBufferProvideChunks(
+    _Inout_ QUIC_RECV_BUFFER* RecvBuffer,
+    _Inout_ CXPLAT_LIST_ENTRY* /* QUIC_RECV_CHUNKS */ Chunks
+    )
+{
+    CXPLAT_DBG_ASSERT(RecvBuffer->RecvMode == QUIC_RECV_BUF_MODE_APP_OWNED);
+    CXPLAT_DBG_ASSERT(!CxPlatListIsEmpty(Chunks));
+
+    uint64_t NewBufferLength = RecvBuffer->VirtualBufferLength;
+    for (CXPLAT_LIST_ENTRY* Link = Chunks->Flink;
+         Link != Chunks;
+         Link = Link->Flink) {
+        QUIC_RECV_CHUNK* Chunk = CXPLAT_CONTAINING_RECORD(Link, QUIC_RECV_CHUNK, Link);
+        NewBufferLength += Chunk->AllocLength;
+    }
+
+    if (NewBufferLength > UINT32_MAX) {
```

Comment:
Is this already ensured at the layer above? If so, we should debug assert and not add branches at multiple layers.

Reply:
And if we can remove this error check, we then simplify the function by no longer needing a return code.

Reply:
We can't validate the total buffer size before because it depends on the current `VirtualBufferLength`, and adding the buffers can be async.
All that we validate earlier is that the list is not empty, and each individual buffer size is non-zero.